amr-wind/core/field_ops.H

```diff
@@ -293,36 +293,109 @@ inline void lincomb(
     int numcomp,
     int nghost)
 {
-    lincomb(dst, a, x, xcomp, b, y, ycomp, dstcomp, numcomp, amrex::IntVect(nghost));
+    lincomb(
+        dst, a, x, xcomp, b, y, ycomp, dstcomp, numcomp,
+        amrex::IntVect(nghost));
 }
 
-/** Set the lower bound for a given scalar field   
+/** Set the lower bound for a given scalar field
  * \ingroup field_ops
- * 
+ *
  * \param [in] min_value Minimum value for a given field
  * \param [in] icomp Component to set the minimum bound
  */
-template<typename FType>
-inline void lower_bound(
-    FType& field,
-    const amrex::Real min_value,
-    const int icomp = 0)
+template <typename FType>
+inline void
+lower_bound(FType& field, const amrex::Real min_value, const int icomp = 0)
 {
-  const auto& repo = field.repo();
-  const int nlevels = repo.num_active_levels();
-  for (int lev=0; lev < nlevels; ++lev) {
-    
-    for (amrex::MFIter mfi(field(lev)); mfi.isValid(); ++mfi) {
-      const auto& bx = mfi.tilebox();
-      const auto& field_arr = field(lev).array(mfi);
+    const auto& repo = field.repo();
+    const int nlevels = repo.num_active_levels();
+    for (int lev = 0; lev < nlevels; ++lev) {
 
-      amrex::ParallelFor(
-        bx, [=] AMREX_GPU_DEVICE(int i, int j, int k) noexcept {
-          field_arr(i,j,k,icomp) = amrex::max(min_value, field_arr(i,j,k,icomp));
-      });
+        for (amrex::MFIter mfi(field(lev)); mfi.isValid(); ++mfi) {
+            const auto& bx = mfi.tilebox();
+            const auto& field_arr = field(lev).array(mfi);
+
+            amrex::ParallelFor(
+                bx, [=] AMREX_GPU_DEVICE(int i, int j, int k) noexcept {
+                    field_arr(i, j, k, icomp) =
+                        amrex::max(min_value, field_arr(i, j, k, icomp));
+                });
+        }
     }
-  }
+}
+
+/** Computes the global maximum of a field from all levels
+ * \ingroup field_ops
+ *
+ * \param [in] field we need to compute its global maximum magnitude
+ */
+template <typename FType>
+inline amrex::Real global_max_magnitude(FType& field)
+{
+    using namespace amrex;
+    const auto& repo = field.repo();
+    const auto ncomp = field.num_comp();
 
+    amrex::Real maxglobal;
+    const int nlevels = repo.num_active_levels();
+    for (int lev = 0; lev < nlevels; ++lev) {
+        amrex::Real maxglobal = 0.0;
```

Comment:
This `maxglobal` is shadowing the outer variable. Has this code been tested?

Reply:
I am working on a unit test now. this is the same idea as in incflo_compute_dt.cpp --> https://github.com/Exawind/amr-wind/blob/569d308a9b788634035236c20143c4a31645d660/amr-wind/incflo_compute_dt.cpp#L52-#L64